### PR TITLE
Use new NA scalar in BooleanArray

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Follow-up on https://github.com/pandas-dev/pandas/pull/29597 and https://github.com/pandas-dev/pandas/pull/29555, now actually using the `pd.NA` scalar in `BooleanArray`.

### CODE REVIEW

File: pandas/core/arrays/boolean.py [hunk 1]
```diff
@@ -61,13 +61,13 @@ class BooleanDtype(ExtensionDtype):
     @property
     def na_value(self) -> "Scalar":
         """
-        BooleanDtype uses :attr:`numpy.nan` as the missing NA value.
```
Comment (TomAugspurger) [inline pandas/core/arrays/boolean.py]: Does this render correctly, or does it need to be `pandas` rather than `pd`?

**Review by TomAugspurger — COMMENTED** [review]
LGTM at a glance. Request for a few more tests / confirmation that these are already testsed:

1. Test to enure that `array([True, False, None, np.nan pd.NA], dtype="boolean")` correctly sanitizes all the NA-like value to be NA.
2. Test in `BooleanArray.__setitem__` ensuring that `arr[0] = np.nan`, etc., always inserts `NA`.

**Comment by jorisvandenbossche** [comment]
One question that comes up here (but the same question applies to string and integer arrays): what "NA value" do we want to use when converting to object dtype numpy arrays? `None` or `pd.NA` ? 

In the initial BooleanArray PR, I used `None` (since `pd.NA` was not there yet), so you get a numpy array like `np.array([True, False, None])`. 
Now we can start using `pd.NA`, which is closer to the pandas representation (and what you would get from iteration or conversion to list `np.array([i for i in arr])`). But on the other hand, `None` can be easier to handle for cases where you need a numpy array (functionality that needs numpy arrays will typically also not recognize or handle correctly  `pd.NA`).

**Comment by TomAugspurger** [comment]
I'm not sure, but my initial preference is for having `pd.NA`, with an option to get other values for NA upon request. I think that means we should have a somewhat standard `to_numpy` method

```python
def to_numpy(self, dtype=object, na_value=pd.NA):
    ... 
```

Then if the user wants None / NaN, they can request it relatively easily.

> do we want to use when converting to object dtype numpy arrays?

What are the user-actions that hit this?

1. `np.array(boolarray, dtype=...)`
2. `boolarray.astype(np_dtype)`
3. ...?

**Comment by jorisvandenbossche** [comment]
Examples where the `pd.NA` in a numpy array gives problems: https://github.com/pandas-dev/pandas/issues/29976 (pyarrow's C conversion code does not know it) and the factorize errors in https://github.com/pandas-dev/pandas/pull/29964 (our cython hashing code for object arrays cannot handle it). 

Now, both are easy to solve (certainly since both can easily be solved on our side; the hashing code can recognize pd.NA and for the conversion to pyarrow we can ensure to use None before passing to pyarrow). But it are examples of how other code can break. 

Still not sure that we should therefore use None as the default in `__array__`, but at least I think it is important to have this `to_numpy(.., na_value=..)`.

File: pandas/core/arrays/boolean.py [hunk 2]
```diff
@@ -281,7 +281,9 @@ def __getitem__(self, item):
             return self._data[item]
         return type(self)(self._data[item], self._mask[item])
 
-    def _coerce_to_ndarray(self, force_bool: bool = False):
+    def _coerce_to_ndarray(
+        self, force_bool: bool = False, na_value: "Scalar" = lib._no_default
```
Comment (TomAugspurger) [inline pandas/core/arrays/boolean.py]: Any reason to prefer `lib_no_default` to just `libmissing.NA` directly?

**Review by TomAugspurger — COMMENTED** [review]
For the "what value to use when converting", I'm not sure that downstream projects not understanding `pd.NA` should drive our decision here. We'll be living with this decision for a while, and projects will have time to adapt.

File: pandas/core/arrays/boolean.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/boolean.py]: Hmm, not sure. I was probably sidetracked by the idea I could not use None as default as that is a valid value as well .. 
(if we want to have this generic / shared with other arrays, using `self._na_value` might be useful, but I don't think we will share this with arrays that don't use pd.NA, so ..)

**Comment by jorisvandenbossche** [comment]
@TomAugspurger updated this

**Comment by TomAugspurger** [comment]
See https://github.com/pandas-dev/pandas/issues/30043 for the CI failures. OK to ignore for now.

File: pandas/core/arrays/boolean.py [hunk 3]
```diff
@@ -483,8 +480,17 @@ def astype(self, dtype, copy=True):
             return IntegerArray(
                 self._data.astype(dtype.numpy_dtype), self._mask.copy(), copy=False
             )
+        # for integer, error if there are missing values
+        if is_integer_dtype(dtype):
+            if self.isna().any():
+                raise ValueError("cannot convert NA to integer")
+        # for float dtype, ensure we use np.nan before casting (numpy cannot
```
Comment (TomAugspurger) [inline pandas/core/arrays/boolean.py]: Pending the discussion in https://github.com/pandas-dev/pandas/issues/30038.

**Review by TomAugspurger — APPROVED** [review]
I think we can move forward with this, despite the ongoing API discussion about `.to_numpy` / `.astype(float)` / `np.asarray(arr)`, since that will be affecting all IntegerArray / StringArray / BooleanArray.

**Comment by TomAugspurger** [comment]
One more note: we'll need to handle reductions like `any` and `all`. They look somewhat buggy on master though.

```python
In [16]: pd.array([True, None])._reduce('all')
Out[16]: True

In [17]: pd.array([True, None])._reduce('all', skipna=False)
Out[17]: True

In [18]: pd.array([False, None])._reduce('all', )
Out[18]: False

In [19]: pd.array([False, None])._reduce('all', skipna=False)
Out[19]: False
```

Do we want to do that here or as a followup?

**Comment by jorisvandenbossche** [comment]
Let's do that as a follow-up, that's a remainder from the initial implementation, it's noted as a to do item in https://github.com/pandas-dev/pandas/issues/29556

**Comment by TomAugspurger** [comment]
Sounds good to me.

**Comment by jorisvandenbossche** [comment]
OK, going to merge this then, so the Kleene PR can then be updated.

**Comment by jorisvandenbossche** [comment]
Once the IntegerArray PR is in, I will also take a look at consolidating both classes

**Comment by TomAugspurger** [comment]
I'll update the Kleene PR now, and the Integer one after if I have a chance.

**Comment by jorisvandenbossche** [comment]
I did a quick PR for the any/all: https://github.com/pandas-dev/pandas/pull/30062

**Pull request merged**